cli/js/40_jupyter.js

```diff
@@ -336,8 +356,90 @@ async function formatInner(obj, raw) {
 
 internals.jupyter = { formatInner };
 
+/**
+class CommMessage(TypedDict):
+    header: dict
+    # typically UUID, must be unique per message
+    msg_id: str
+    msg_type: str
+    parent_header: dict
+    metadata: dict
+    content: <custom payload>
+    buffers: list[memoryview]
+
+((async) => {
+  const data = await Deno.jupyter.comms.recv("1234-5678");
+})();
+((async) => {
+  const data = await Deno.jupyter.comms.recv("1234-5678");
+})();
+
+const comm = await Deno.jupyter.comms.open("1234-5678");
+const data = await comm.recv();
+
+const data = await Deno.jupyter.comms.recv("1234-5678");
+
+c = Comm("1234-5678")
+
+c.on("update", data => {
+    console.log(data);
+    Deno.jupyter.broadcast(...);
+});
+
+
+{
+    msg_type: "comm_msg",
+    content: {
+        comm_id: "1234-5678",
+        data: {
+
+        }
+    }
+}
+*/
+
 function enableJupyter() {
-  const { op_jupyter_broadcast } = core.ops;
+  const { op_jupyter_broadcast, op_jupyter_comm_recv, op_jupyter_comm_open } =
+    core.ops;
+
+  function comm(commId, targetName, msgCallback) {
+    op_jupyter_comm_open(commId, targetName);
+
+    let closed = false;
+
+    // TODO(bartlomieju): return something, so we can close this comm.
+    (async () => {
+      while (true) {
+        const [data, buffers] = await op_jupyter_comm_recv(commId);
+
+        if (!data) {
+          closed = true;
+          break;
+        }
+
+        msgCallback?.({
+          ...data,
+          buffers,
+        });
+      }
+    })();
+
+    return {
+      close() {
+        if (closed) {
+          return;
+        }
+
+        closed = true;
+      },
+      send(data) {
+        return broadcast("comm_msg", {
+          comm_id: commId,
+          data: data,
+        });
+      },
```

Comment:
In order to send binary data to widgets, we'll also want to allow for buffers here:

```suggestion
      send(data, buffers = []) {
        return broadcast("comm_msg", {
          comm_id: commId,
          data: data,
        }, { buffers });
      },
```

Reply:
Also, I think the recv callback should mirror whatever we have here for sending. So maybe rather than `{ ...data, buffers }` we can just have a function with two args:

```js
let comm = Deno.jupyter.comm(commId, (data, buffers) => {

})

comm.send(data, buffers);
```

Reply:
Bikeshedding the API... right now you still need to open the comm with broadcast separately ("comm_open") message.

```js
let comm = await Deno.jupyter.commOpen(commId, {
  targetName: "jupyter.widgets",
  data: { /* anywidget stuff */ },
  metadata: { version: "2.1.0" },
});

// either only allow one callback or manage a set of them
comm.on((data, buffers) => {

});

comm.close();

comm.send(data, buffers);
```

or something more webby (e.g., EventTarget, [`AudioContext`](https://developer.mozilla.org/en-US/docs/Web/API/AudioContext)):

```js
let comm = new Deno.jupyter.Comm(commId, {
  targetName: "jupyter.widgets",
  data: { /* anywidget stuff */ },
  metadata: { version: "2.1.0" },
});

comm.addEventListener("comm_msg", (event) => {
  let { data, buffers } = event.detail;
});

comm.removeEventListener("comm_msg", handler);

comm.dispatchEvent(
  new CustomEvent("comm_msg", { detail: { data, buffers } })
);

comm.close();
```